kong/core/router.lua

```diff
@@ -264,48 +312,100 @@ end
 
 do
   local matchers = {
-    [MATCH_RULES.HOST] = function(api_t, _, _, host)
+    [MATCH_RULES.HOST] = function(api_t, ctx)
+      local host = ctx.wildcard_host or ctx.host
+
       if api_t.hosts[host] then
         return true
       end
     end,
 
-    [MATCH_RULES.URI] = function(api_t, _, uri)
-      if api_t.uris[uri] then
-        if api_t.strip_uri then
-          api_t.strip_uri_regex = api_t.uris[uri].strip_regex
-        end
+    [MATCH_RULES.URI] = function(api_t, ctx)
+      do
+        local uri     = ctx.uri_regex or ctx.uri_prefix or ctx.uri
+        local regex_t = api_t.uris[uri]
+
+        if regex_t then
+          if api_t.strip_uri or regex_t.has_captures then
+            local m, err = re_match(ctx.uri, regex_t.strip_regex, "ajo")
+            if err then
+              log(ERR, "could not evaluate URI prefix/regex: ", err)
+              return
+            end
 
-        return true
-      end
+            if m then
+              if m.stripped_uri then
+                ctx.stripped_uri = "/" .. m.stripped_uri
+                -- remove the stripped_uri group
+                m[#m]            = nil
+                m.stripped_uri   = nil
+              end
 
-      for i = 1, #api_t.uris_prefixes_regexes do
-        local from, _, err = re_find(uri, api_t.uris_prefixes_regexes[i].regex, "ajo")
-        if err then
-          log(ERR, "could not search for URI prefix: ", err)
-          return
-        end
+              if regex_t.has_captures then
+                m[0] = nil
```

Comment:
There would be, but only if we also expose a match from a plain/prefix URI as well, otherwise the behavior is inconsistent between the two possibilities. Since we do not (for now) expose it on a plain match, we do not expose it on a regexp match either.